Add support for the Web Extension commands API and events.

xeenon opened the pull request:
#### 5b25309d1e6f03442f6e5a8638f43844958c0da5
<pre>
Add support for the Web Extension commands API and events.
<a href="https://bugs.webkit.org/show_bug.cgi?id=260157">https://bugs.webkit.org/show_bug.cgi?id=260157</a>
<a href="https://rdar.apple.com/problem/114823250">rdar://problem/114823250</a>

Reviewed by Brent Fulgham.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_cannot_be_forward_declared):
* Source/WebKit/Shared/Extensions/WebExtensionCommandParameters.h: Added.
* Source/WebKit/Shared/Extensions/WebExtensionCommandParameters.serialization.in: Added.
* Source/WebKit/Shared/Extensions/WebExtensionEventListenerType.h: Added CommandsOnChanged and sorted.
* Source/WebKit/Shared/Extensions/WebExtensionEventListenerType.serialization.in: Ditto.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPICommandsCocoa.mm: Added.
(WebKit::WebExtensionContext::commandsGetAll):
(WebKit::WebExtensionContext::fireCommandEventIfNeeded):
(WebKit::WebExtensionContext::fireCommandChangedEventIfNeeded):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::hasAction): Added.
(WebKit::WebExtension::hasBrowserAction): Added.
(WebKit::WebExtension::hasPageAction): Added.
(WebKit::WebExtension::populateCommandsIfNeeded): Added action command.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCommandCocoa.mm:
(WebKit::WebExtensionCommand::dispatchChangedEventSoonIfNeeded): Added.
(WebKit::WebExtensionCommand::setActivationKey): Call dispatchChangedEventSoonIfNeeded().
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::performCommand): Call performAction() and fireCommandEventIfNeeded().
* Source/WebKit/UIProcess/Extensions/WebExtension.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionCommand.cpp:
(WebKit::WebExtensionCommand::isActionCommand const): Added.
(WebKit::WebExtensionCommand::parameters const): Added.
* Source/WebKit/UIProcess/Extensions/WebExtensionCommand.h:
(WebKit::WebExtensionCommand::setModifierFlags):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPICommandsCocoa.mm: Added.
(WebKit::toAPI):
(WebKit::WebExtensionAPICommands::getAll):
(WebKit::WebExtensionAPICommands::onCommand):
(WebKit::WebExtensionAPICommands::onChanged):
(WebKit::WebExtensionContextProxy::dispatchCommandsCommandEvent):
(WebKit::WebExtensionContextProxy::dispatchCommandsChangedEvent):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPINamespaceCocoa.mm:
(WebKit::WebExtensionAPINamespace::isPropertyAllowed):
(WebKit::WebExtensionAPINamespace::commands):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPICommands.h: Added.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPINamespace.h:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPICommands.idl: Added.
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPINamespace.idl: Added commands.
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.messages.in:
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtension.mm:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPICommands.mm: Added.
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionContext.mm:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/270523@main">https://commits.webkit.org/270523@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/25d17e1a58094ea4e104bef464654894248d7a0f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25692 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4298 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/26976 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/27793 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23531 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/25976 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6050 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/1733 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/23656 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/25941 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3211 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22143 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28373 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/25856 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/2836 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23098 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29170 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/23447 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23464 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27029 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/2855 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1091 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4236 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6168 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3305 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3171 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPICommandsCocoa.mm:
```diff
@@ -0,0 +1,73 @@
+    results.reserveInitialCapacity(commands().size());
+
+    for (auto& command : commands())
+        results.append(command->parameters());
```
Delete lines 46-50 and replace with:
```
auto results = WTF::map(commands(), [](auto&& command) -> WebExtensionComandParameters {
    return command->parameters();
};
```

brentfulgham commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1701,6 +1701,35 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+        return m_commands;
+
+    for (auto& data : extension().commands())
+        m_commands.append(WebExtensionCommand::create(*this, data));
```
Since `m_commands` is uninitialized here, and we don't hint at the size, it could be expensive to loop and append (if there are a lot of commands -- maybe there never are).

Still it's simple to write the code this way and avoid the risk entirely:

I think `extension().commands()` returns a copy, so you can use a move operator into the lambda. If that's wrong, just make it a reference:

```
m_commands = WTF::map(extension().commands(), [&](auto&& data) -> Ref<WebExtensionCommand> {
    return WebExtensionCommand::create(*this, data);
|);
```

brentfulgham reviewed (approved):
I had a couple of suggestions for our preferred way to initialize vectors to avoid performance issues, but otherwise I think this looks good. r=me!

xeenon committed:
Add support for the Web Extension commands API and events.
https://bugs.webkit.org/show_bug.cgi?id=260157
rdar://problem/114823250

Reviewed by Brent Fulgham.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_cannot_be_forward_declared):
* Source/WebKit/Shared/Extensions/WebExtensionCommandParameters.h: Added.
* Source/WebKit/Shared/Extensions/WebExtensionCommandParameters.serialization.in: Added.
* Source/WebKit/Shared/Extensions/WebExtensionEventListenerType.h: Added CommandsOnChanged and sorted.
* Source/WebKit/Shared/Extensions/WebExtensionEventListenerType.serialization.in: Ditto.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPICommandsCocoa.mm: Added.
(WebKit::WebExtensionContext::commandsGetAll):
(WebKit::WebExtensionContext::fireCommandEventIfNeeded):
(WebKit::WebExtensionContext::fireCommandChangedEventIfNeeded):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::hasAction): Added.
(WebKit::WebExtension::hasBrowserAction): Added.
(WebKit::WebExtension::hasPageAction): Added.
(WebKit::WebExtension::populateCommandsIfNeeded): Added action command.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCommandCocoa.mm:
(WebKit::WebExtensionCommand::dispatchChangedEventSoonIfNeeded): Added.
(WebKit::WebExtensionCommand::setActivationKey): Call dispatchChangedEventSoonIfNeeded().
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::performCommand): Call performAction() and fireCommandEventIfNeeded().
* Source/WebKit/UIProcess/Extensions/WebExtension.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionCommand.cpp:
(WebKit::WebExtensionCommand::isActionCommand const): Added.
(WebKit::WebExtensionCommand::parameters const): Added.
* Source/WebKit/UIProcess/Extensions/WebExtensionCommand.h:
(WebKit::WebExtensionCommand::setModifierFlags):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPICommandsCocoa.mm: Added.
(WebKit::toAPI):
(WebKit::WebExtensionAPICommands::getAll):
(WebKit::WebExtensionAPICommands::onCommand):
(WebKit::WebExtensionAPICommands::onChanged):
(WebKit::WebExtensionContextProxy::dispatchCommandsCommandEvent):
(WebKit::WebExtensionContextProxy::dispatchCommandsChangedEvent):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPINamespaceCocoa.mm:
(WebKit::WebExtensionAPINamespace::isPropertyAllowed):
(WebKit::WebExtensionAPINamespace::commands):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPICommands.h: Added.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPINamespace.h:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPICommands.idl: Added.
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPINamespace.idl: Added commands.
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.messages.in:
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtension.mm:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPICommands.mm: Added.
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionContext.mm:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/270523@main
